File: 1650-lowest-common-ancestor-of-a-binary-tree-iii/1650-lowest-common-ancestor-of-a-binary-tree-iii.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    Node* left;
    Node* right;
    Node* parent;
};
*/

class Solution {
public:
    Node* lowestCommonAncestor(Node* p, Node * q) {
        unordered_set<Node*> set;
        
        
        while(p != nullptr)
        {
            set.insert(p);
            p = p->parent;
        }
        
        while(q != nullptr)
        {
            if(set.find(q) != set.end()) return q;
            q = q->parent;
        }
        
        return p;
    }
};

File: 1650-lowest-common-ancestor-of-a-binary-tree-iii/1650-lowest-common-ancestor-of-a-binary-tree-iii.py
"""
# Definition for a Node.
class Node:
    def __init__(self, val):
        self.val = val
        self.left = None
        self.right = None
        self.parent = None
"""

class Solution:
    def lowestCommonAncestor(self, p: 'Node', q: 'Node') -> 'Node':
        
        
        def dfs(x, y):
            if x.val == y.val:
                return x
            
            if x.parent == None:
                return dfs(q, y.parent)
            elif y.parent == None:
                return dfs(x.parent, p)
            else:
                return dfs(x.parent, y.parent)
        
        return dfs(p, q)